### PR TITLE
cs_volume: add volumes extraction and upload features

### PR DESCRIPTION
**Pull request by dpassante**
##### SUMMARY

Add volumes extraction and upload features that could be useful on cross-cluster or cross-zone migration scenarios.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

cs_volume

##### ADDITIONAL INFORMATION

```paste below
- name: Extract DATA volume to make it downloadable
  cs_volume:
    api_region: "{{ source.region }}"
    state: extracted
    name: "{{ data_volume }}"
  register: data_vol_out
  delegate_to: localhost

- name: Download DATA volume into destination
  cs_volume:
    api_region: "{{ destination.region }}"
    state: uploaded
    name: "DATA-{{ inventory_hostname }}"
    format: VHD
    url: "{{ data_vol_out.url }}"
  delegate_to: localhost
```
@resmo I implemented these features in the `cs_volume` module to quickly have something functional. It's still a work in progress because of the lack of examples and unit tests.
Let me know if you'd prefer to see these features implemented on own modules.

### CODE REVIEW

**Comment by resmo** [comment]
I am totally fine to implement this in cs_volume.

File: lib/ansible/modules/cloud/cloudstack/cs_volume.py [hunk 1]
```diff
@@ -98,7 +98,7 @@
     description:
       - State of the volume.
```
Comment (resmo) [inline lib/ansible/modules/cloud/cloudstack/cs_volume.py:97]:
~~~diff
    description:
      - State of the volume.
+    - The choices I(extracted) and I(uploaded) were added in version 2.8. 
~~~

File: lib/ansible/modules/cloud/cloudstack/cs_volume.py [hunk 2]
```diff
@@ -110,6 +110,28 @@
...
+  url:
+    description:
+      - URL to which the volume would be extracted on I(state=extracted)
+      - or the URL where to download the volume on I(state=uploaded).
+      - Only considered if I(state) is C(extracted) or C(uploaded).
+    type: str
+    version_added: '2.8'
+  mode:
+    description:
+      - Mode for the template extraction.
```
Comment (resmo) [inline lib/ansible/modules/cloud/cloudstack/cs_volume.py]: s/template/volume ?

File: lib/ansible/modules/cloud/cloudstack/cs_volume.py (hunk 1 above)
Comment (dpassante) [reply lib/ansible/modules/cloud/cloudstack/cs_volume.py:97]: OK! with C() as these are option values ;-)

File: lib/ansible/modules/cloud/cloudstack/cs_volume.py (hunk 2 above)
Comment (dpassante) [reply lib/ansible/modules/cloud/cloudstack/cs_volume.py]: whoops ! copy/pasta

File: lib/ansible/modules/cloud/cloudstack/cs_volume.py [hunk 3]
```diff
@@ -398,6 +431,57 @@ def update_volume(self, volume):
...
+                'domainid': self.get_domain(key='id'),
+                'projectid': self.get_project(key='id'),
+                'zoneid': self.get_zone(key='id'),
+                'format': self.module.params.get('format'),
+                'url': self.module.params.get('url'),
+                'diskofferingid': disk_offering_id,
+            }
+            if not self.module.check_mode:
+                res = self.query_api('uploadVolume', **args)
+                if 'errortext' in res:
```
Comment (resmo) [inline lib/ansible/modules/cloud/cloudstack/cs_volume.py]: deprecated code

**Comment by dpassante** [comment]
ready_for_review

I noticed `device_id` param was missing on the doc. I didn't added it yet to avoid having problems with validate-module.

**Review by resmo — APPROVED** [review]
perfect job!

shipit

**Comment by dpassante** [comment]
@resmo Thanks René !

**Pull request merged**